https://drive.google.com/drive/folders/1l7Fxjyr1NXgoNCby2-ow-dVmlkywixKz?usp=sharing

In [1]:
import numpy as np
import pandas
import seaborn as sns
import matplotlib.pyplot as plt
#importing the necessary libaries
import os

/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


In [2]:
# from google.colab import drive
# drive.mount('/content/drive')

In [10]:
# !tar -xzvf "/lfwdataset/lfw (1).tgz" -C "/lfw"
# Used to unzip the file

tar (child): /lfwdataset/lfw (1).tgz: Cannot open: No such file or directory
tar (child): Error is not recoverable: exiting now
tar: Child returned status 2
tar: Error is not recoverable: exiting now


In [14]:
MINIMUM_SIZE = 2
# path = '/content/drive/MyDrive/Assignment2/lfw/lfw'
path = '/kaggle/input/lfwdataset1/lfw/lfw'

# file_count = 0
count=0
for entry in os.scandir(path):
    file_count = 0
    for fil in os.scandir(entry):
        if fil.is_file():
            file_count += 1
        if file_count == MINIMUM_SIZE:
            count+=1

# loop through the folders and files to find people with more than 2 images
print("Number of people with 2 or more images : ", count)

Number of people with 2 or more images :  1680


In [15]:
count=0
for entry in os.scandir(path):
    count+=1
print("Number of folders/distinct people : ", count)

# The number of distinct people

Number of folders/distinct people :  603645


The dataset contains 1680 distinct people. Here images for only 1680 people are used. These are divided as follows -

Training data - 560
Validation data - 140
Test data - 100

In [17]:
from PIL import Image
import os
 
# path = '/content/drive/MyDrive/Assignment2/lfw/lfw/'
path = '/kaggle/input/lfwdataset1/lfw/lfw'

count = 0

for entry in os.listdir(path):
    entry_path = os.path.join(path, entry)

    # Check if the entry is a main folder
    if os.path.isdir(entry_path):

        # Iterate through the files in the sub-folders
        for fil in os.listdir(entry_path):
            file_path = os.path.join(entry_path, fil)

            if file_path.endswith(('.jpg', '.jpeg', '.png', '.gif')):
                print(f"Image: {file_path}")

                # Open the image and get its size
                img = Image.open(file_path)
                width, height = img.size
                print(f"Size: {width} x {height}")
                count+=1
        if(count==10):
            break
    break

print("All images are of size 250*250")

#reference - chatgpt

Image: /kaggle/input/lfwdataset1/lfw/lfw/Tyler_Hamilton/Tyler_Hamilton_0002.jpg
Size: 250 x 250
Image: /kaggle/input/lfwdataset1/lfw/lfw/Tyler_Hamilton/Tyler_Hamilton_0001.jpg
Size: 250 x 250
All images are of size 250*250


The code below divides the images into train, validation and test set.

In [19]:
# from PIL import Image
# import os

# # Define the output size for the resized images
# output_size = (200, 200)  # Change this to the desired size

# input_path = '/content/drive/MyDrive/Assignment2/lfw/lfw/'

# file_count = 0

# # Define the thresholds for the split
# validation_threshold = 560  # Number of persons for training
# test_threshold = 700  # Number of persons for validation

# # Code below same as above cell
# for entry in os.listdir(input_path):
#     entry_path = os.path.join(input_path, entry)

#     if os.path.isdir(entry_path):
#         # Create a subdirectory to store the cropped and resized images
#         if file_count >= validation_threshold and file_count < test_threshold:
#             output_path = '/content/drive/MyDrive/Assignment2/lfw/validation_set/'
#         elif file_count >= test_threshold and file_count<=800:
#             output_path = '/content/drive/MyDrive/Assignment2/lfw/test_set/'
#         elif file_count<validation_threshold and file_count>=0:
#           output_path = '/content/drive/MyDrive/Assignment2/lfw/train_set/'
#         else:
#           break

#         output_dir = os.path.join(output_path, entry)
#         os.makedirs(output_dir, exist_ok=True)

#         for fil in os.listdir(entry_path):
#             file_path = os.path.join(entry_path, fil)

#             if file_path.endswith(('.jpg', '.jpeg', '.png', '.gif')):
#                 # Open the image
#                 img = Image.open(file_path)

#                 # Resize the image to the specified output size
#                 img = img.resize(output_size)

#                 # Save the resized image to the output directory
#                 output_file_path = os.path.join(output_dir, fil)
#                 img.save(output_file_path)

#         file_count += 1

# print(f"All images are resized to {output_size[0]}x{output_size[1]} and saved in the directory.")

In [20]:
import os
import random

# path to the dataset
# lfw_dir = '/content/drive/MyDrive/Assignment2/lfw/train_set'
lfw_dir = '/kaggle/input/lfwdataset1/lfw/train_set'


# Get a list of subfolder for each person
person_folders = [os.path.join(lfw_dir, folder) for folder in os.listdir(lfw_dir)]

# Initialize lists to store positive and negative pairs
positive_pairs = []
negative_pairs = []

# Generate pairs of similar and dissimilar images
for folder in person_folders:
    images = os.listdir(folder)
    if len(images) >= 2:
        # Generate positive pairs
        positive_pairs.extend([(os.path.join(folder, images[i]), os.path.join(folder, images[j])) for i in range(len(images)) for j in range(i + 1, len(images))])

# Generate negative pairs
for i in range(len(person_folders)):
    for j in range(i + 1, len(person_folders)):
        folder1_images = os.listdir(person_folders[i])
        folder2_images = os.listdir(person_folders[j])
        if folder1_images and folder2_images:
            negative_pairs.append((os.path.join(person_folders[i], random.choice(folder1_images)), os.path.join(person_folders[j], random.choice(folder2_images))))


In [22]:
# Assign labels to positive and negative pairs
positive_labels = [1] * len(positive_pairs)
negative_labels = [0] * len(negative_pairs)

print(negative_pairs[0])
print(len(positive_pairs), len(negative_pairs))
# Combine positive and negative pairs and labels
paired_images = positive_pairs + negative_pairs
labels = positive_labels + negative_labels

('/kaggle/input/lfwdataset1/lfw/train_set/Xanana_Gusmao/Xanana_Gusmao_0001.jpg', '/kaggle/input/lfwdataset1/lfw/train_set/Kai-Uwe_Ricke/Kai-Uwe_Ricke_0001.jpg')
20221 156520


In [23]:
import torch
import torch.nn as nn
import torch.optim as optim
import torchvision
from torch.utils.data import Dataset, DataLoader
import torchvision.transforms as transforms

#importing necessary libraries for model architecture and training

In [24]:
import torch
import torch.nn as nn
import torchvision.models as models

# Define the Siamese Network class
class SiameseNetwork(nn.Module):
    def __init__(self):
        super(SiameseNetwork, self).__init__()

        # Load the pre-trained ResNet-50 model
        self.resnet = models.resnet50(pretrained=True)

        # Remove the final classification layer
        self.resnet = nn.Sequential(*list(self.resnet.children())[:-1])

        # Define the fully connected layers for the Siamese network namely 2 dense fully connected layers
        self.fc = nn.Sequential(
            nn.Linear(2048, 512),
            nn.ReLU(),
            nn.Linear(512, 256)
        )

    def forward_one(self, x):
        x = self.resnet(x)
        x = x.view(x.size(0), -1)
        x = self.fc(x)
        return x

    def forward(self, input1, input2):
        output1 = self.forward_one(input1)
        output2 = self.forward_one(input2)
        return output1, output2


In [26]:
import torch.nn.functional as F

class ContrastiveLossWithRegularization(nn.Module):
    def __init__(self, margin=2.5, regularization_strength=0.01):
        super(ContrastiveLossWithRegularization, self).__init__()
        self.margin = margin
        self.regularization_strength = regularization_strength

    def forward(self, output1, output2, label, model):
        euclidean_distance = F.pairwise_distance(output1, output2, keepdim=True)
        loss_similarity = torch.mean((1 - label) * torch.pow(euclidean_distance, 2))
        loss_dissimilarity = torch.mean(label * torch.pow(torch.clamp(self.margin - euclidean_distance, min=0.0), 2))

        # Calculate L2 regularization term
        l2_regularization = 0.0
        for param in model.parameters():
            l2_regularization += torch.norm(param, p=2)

        # Combine the contrastive loss and regularization
        total_loss = loss_similarity + loss_dissimilarity + self.regularization_strength * l2_regularization

        return total_loss


In [27]:
import torch
from torch.utils.data import Dataset
from torchvision import transforms
from PIL import Image

import random
from PIL import Image

#Creating a custom dataset for loading the data into the Siamese network

class SiameseDataset(Dataset):
    def __init__(self, paired_images, labels, transforms):
        self.paired_images = paired_images
        self.labels = labels
        self.transforms = transforms

    def __len__(self):
        return len(self.paired_images)

    def __getitem__(self, idx):
        image1_path, image2_path = self.paired_images[idx]
        label = self.labels[idx]

        # Load the images
        image1 = Image.open(image1_path)
        image2 = Image.open(image2_path)

        # Apply a transformation
        if self.transforms:
            image1 = self.transforms(image1)
            image2 = self.transforms(image2)

        return image1, image2, label


In [28]:
import random
import torch.nn.functional as F

In [29]:
from torchvision import transforms

# Using image augmentation before the images is passed to the Siamese Network
transform_list = transforms.Compose([
    transforms.RandomHorizontalFlip(),  # Random horizontal flip
    transforms.RandomRotation(15),  # Random rotation by 15 degrees
    transforms.ColorJitter(brightness=0.2, contrast=0.2, saturation=0.2, hue=0.2),  # Random color adjustments
    transforms.ToTensor(),  # Convert to a tensor
])


In [30]:
# train_path = '/content/drive/MyDrive/Assignment2/lfw/train_set'
# validation_path = '/content/drive/MyDrive/Assignment2/lfw/validation_set'
# test_path = '/content/drive/MyDrive/Assignment2/lfw/test_set'


# Check if CUDA is available
if torch.cuda.is_available():
    device = torch.device("cuda")
else:
    device = torch.device("cpu")

train_dataset = SiameseDataset(paired_images, labels, transforms = transform_list)

batch_size = 32
train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)

In [31]:
torch.cuda.empty_cache()
print(torch.cuda.memory_summary(device=device, abbreviated=False))

|===========================================================================|
|                  PyTorch CUDA memory summary, device ID 0                 |
|---------------------------------------------------------------------------|
|            CUDA OOMs: 0            |        cudaMalloc retries: 0         |
|===========================================================================|
|        Metric         | Cur Usage  | Peak Usage | Tot Alloc  | Tot Freed  |
|---------------------------------------------------------------------------|
| Allocated memory      |      0 B   |      0 B   |      0 B   |      0 B   |
|       from large pool |      0 B   |      0 B   |      0 B   |      0 B   |
|       from small pool |      0 B   |      0 B   |      0 B   |      0 B   |
|---------------------------------------------------------------------------|
| Active memory         |      0 B   |      0 B   |      0 B   |      0 B   |
|       from large pool |      0 B   |      0 B   |      0 B   |

In [32]:
import torch.optim as optim

margin = 2.5  # Adjust margin as needed
model = SiameseNetwork().to(device)
criterion = ContrastiveLossWithRegularization(margin)

# Optimizer 1: Stochastic Gradient Descent (SGD)
optimizer_sgd = optim.SGD(model.parameters(), lr=0.005, momentum=0.9, weight_decay=1e-4)

# Optimizer 2: Adam
optimizer_adam = optim.Adam(model.parameters(), lr=0.0005, weight_decay=1e-4)


/opt/conda/lib/python3.10/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/opt/conda/lib/python3.10/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=ResNet50_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet50_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)
Downloading: "https://download.pytorch.org/models/resnet50-0676ba61.pth" to /root/.cache/torch/hub/checkpoints/resnet50-0676ba61.pth
100%|██████████| 97.8M/97.8M [00:00<00:00, 248MB/s]


In [ ]:
# # Training loop
# num_epochs = 1  # Adjust as needed
# for epoch in range(num_epochs):
#     for i, data in enumerate(train_loader, 0):
#         input1, input2, label = data
#         input1, input2, label = input1.to(device), input2.to(device), label.to(device)
#         optimizer_adam.zero_grad()
#         output1, output2 = model(input1, input2)
#         loss = criterion(output1, output2, label, model)
#         loss.backward()
#         optimizer_adam.step()

#         if (i + 1) % 10 == 0:
#             print(f"Epoch [{epoch + 1}/{num_epochs}] Batch [{i + 1}/{len(train_loader)}] Loss: {loss.item():.4f}")

# # Adam optimizer with no learning rate scheduler

LR Scheduler: https://d2l.ai/chapter_optimization/lr-scheduler.html

Using 2 learning schedulers - Step and Cosine Annealing and 2 optimizers - SGD and ADAM.

In [33]:
from torch.optim.lr_scheduler import CosineAnnealingLR
num_epochs = 1

# Create a CosineAnnealingLR scheduler
scheduler1 = CosineAnnealingLR(optimizer_adam, T_max=num_epochs)

for epoch in range(num_epochs):
    for i, data in enumerate(train_loader, 0):
        input1, input2, label = data
        input1, input2, label = input1.to(device), input2.to(device), label.to(device)
        optimizer_adam.zero_grad()
        output1, output2 = model(input1, input2)
        loss = criterion(output1, output2, label, model)
        loss.backward()
        optimizer_adam.step()

        # Update the learning rate after each optimizer step
        scheduler1.step()

        if (i + 1) % 10 == 0:
            print(f"Epoch [{epoch + 1}/{num_epochs}] Batch [{i + 1}/{len(train_loader)}] Loss: {loss.item():.4f}")


Epoch [1/1] Batch [10/5524] Loss: 8.9178
Epoch [1/1] Batch [20/5524] Loss: 9.5055
Epoch [1/1] Batch [30/5524] Loss: 8.6697
Epoch [1/1] Batch [40/5524] Loss: 8.7592
Epoch [1/1] Batch [50/5524] Loss: 9.3062
Epoch [1/1] Batch [60/5524] Loss: 8.9356
Epoch [1/1] Batch [70/5524] Loss: 9.4482
Epoch [1/1] Batch [80/5524] Loss: 8.6295
Epoch [1/1] Batch [90/5524] Loss: 8.9881
Epoch [1/1] Batch [100/5524] Loss: 8.6265
Epoch [1/1] Batch [110/5524] Loss: 9.0295
Epoch [1/1] Batch [120/5524] Loss: 8.2396
Epoch [1/1] Batch [130/5524] Loss: 8.1142
Epoch [1/1] Batch [140/5524] Loss: 8.1545
Epoch [1/1] Batch [150/5524] Loss: 8.4778
Epoch [1/1] Batch [160/5524] Loss: 7.9510
Epoch [1/1] Batch [170/5524] Loss: 8.7907
Epoch [1/1] Batch [180/5524] Loss: 8.0745
Epoch [1/1] Batch [190/5524] Loss: 7.8371
Epoch [1/1] Batch [200/5524] Loss: 7.7365
Epoch [1/1] Batch [210/5524] Loss: 7.3097
Epoch [1/1] Batch [220/5524] Loss: 7.2163
Epoch [1/1] Batch [230/5524] Loss: 7.4405
Epoch [1/1] Batch [240/5524] Loss: 7.2293
E

Code below resizes the custom images to desired size

In [ ]:
# from PIL import Image
# import os

# # Define the output size for the resized images
# output_size = (200, 200)  # Change this to the desired size

# input_path = '/content/drive/MyDrive/Assignment2/lfw/Custom/'
# output_path = '/content/drive/MyDrive/Assignment2/lfw/Custom1/'

# file_count = 0

# for entry in os.listdir(input_path):
#     entry_path = os.path.join(input_path, entry)

#     if os.path.isdir(entry_path):
#         # Create a subdirectory to store the cropped and resized images

#         output_dir = os.path.join(output_path, entry)
#         os.makedirs(output_dir, exist_ok=True)

#         for fil in os.listdir(entry_path):
#             file_path = os.path.join(entry_path, fil)

#             if file_path.endswith(('.jpg', '.jpeg', '.png', '.gif')):
#                 # Open the image
#                 img = Image.open(file_path)

#                 # Resize the image to the specified output size
#                 img = img.resize(output_size)

#                 # Save the resized image to the output directory
#                 output_file_path = os.path.join(output_dir, fil)
#                 img.save(output_file_path)

#         file_count += 1

# print(f"All images are resized to {output_size[0]}x{output_size[1]} and saved in the directory.")

# # Resizing the custom images

Creating a validation and test dataloader

In [34]:
import os
import random

# path to the dataset
val_dir = '/kaggle/input/lfwdataset1/lfw/validation_set'
# val_dir = '/content/drive/MyDrive/Assignment2/lfw/validation_set'

# Get a list of subfolder for each person
person_folders = [os.path.join(val_dir, folder) for folder in os.listdir(val_dir)]

# Initialize lists to store positive and negative pairs
positive_pairsv = []
negative_pairsv = []

# Generate pairs of similar and dissimilar images
for folder in person_folders:
    images = os.listdir(folder)
    if len(images) >= 2:
        # Generate positive pairs
        positive_pairsv.extend([(os.path.join(folder, images[i]), os.path.join(folder, images[j])) for i in range(len(images)) for j in range(i + 1, len(images))])

# Generate negative pairs
for i in range(len(person_folders)):
    for j in range(i + 1, len(person_folders)):
        folder1_images = os.listdir(person_folders[i])
        folder2_images = os.listdir(person_folders[j])
        if folder1_images and folder2_images:
            negative_pairsv.append((os.path.join(person_folders[i], random.choice(folder1_images)), os.path.join(person_folders[j], random.choice(folder2_images))))


In [35]:
import os
import random

# path to the dataset
# test_dir = '/content/drive/MyDrive/Assignment2/lfw/test_set'
test_dir = '/kaggle/input/lfwdataset1/lfw/test_set'


# Get a list of subfolder for each person
person_folders = [os.path.join(test_dir, folder) for folder in os.listdir(test_dir)]

# Initialize lists to store positive and negative pairs
positive_pairst = []
negative_pairst = []

# Generate pairs of similar and dissimilar images
for folder in person_folders:
    images = os.listdir(folder)
    if len(images) >= 2:
        # Generate positive pairs
        positive_pairst.extend([(os.path.join(folder, images[i]), os.path.join(folder, images[j])) for i in range(len(images)) for j in range(i + 1, len(images))])

# Generate negative pairs
for i in range(len(person_folders)):
    for j in range(i + 1, len(person_folders)):
        folder1_images = os.listdir(person_folders[i])
        folder2_images = os.listdir(person_folders[j])
        if folder1_images and folder2_images:
            negative_pairst.append((os.path.join(person_folders[i], random.choice(folder1_images)), os.path.join(person_folders[j], random.choice(folder2_images))))


In [36]:
# Assign labels to positive and negative pairs
positive_labelsv = [1] * len(positive_pairsv)
negative_labelsv = [0] * len(negative_pairsv)

print(negative_pairsv[0])
print(len(positive_pairsv), len(negative_pairsv))
# Combine positive and negative pairs and labels
validation_paired_images = positive_pairsv + negative_pairsv
validation_labels = positive_labelsv + negative_labelsv

('/kaggle/input/lfwdataset1/lfw/validation_set/Svend_Robinson/Svend_Robinson_0001.jpg', '/kaggle/input/lfwdataset1/lfw/validation_set/Costas_Simitis/Costas_Simitis_0005.jpg')
8599 9730


In [37]:
# Assign labels to positive and negative pairs
positive_labelst = [1] * len(positive_pairst)
negative_labelst = [0] * len(negative_pairst)

print(negative_pairst[0])
print(len(positive_pairst), len(negative_pairst))
# Combine positive and negative pairs and labels
test_paired_images = positive_pairst + negative_pairst
test_labels = positive_labelst + negative_labelst

('/kaggle/input/lfwdataset1/lfw/test_set/Coco_dEste/Coco_dEste_0001.jpg', '/kaggle/input/lfwdataset1/lfw/test_set/Barbara_Bodine/Barbara_Bodine_0001.jpg')
2467 5050


In [41]:
import os
import random

# path to the dataset
# c_dir = '/content/drive/MyDrive/Assignment2/lfw/Custom1'
c_dir = '/kaggle/input/lfwdataset1/lfw/Custom1'

# Get a list of subfolder for each person
person_folders = [os.path.join(c_dir, folder) for folder in os.listdir(c_dir)]

# Initialize lists to store positive and negative pairs
positive_pairsc = []
negative_pairsc = []

# Generate pairs of similar and dissimilar images
for folder in person_folders:
    images = os.listdir(folder)
    if len(images) >= 2:
        # Generate positive pairs
        positive_pairsc.extend([(os.path.join(folder, images[i]), os.path.join(folder, images[j])) for i in range(len(images)) for j in range(i + 1, len(images))])

# Generate negative pairs
for i in range(len(person_folders)):
    for j in range(i + 1, len(person_folders)):
        folder1_images = os.listdir(person_folders[i])
        folder2_images = os.listdir(person_folders[j])
        if folder1_images and folder2_images:
            negative_pairsc.append((os.path.join(person_folders[i], random.choice(folder1_images)), os.path.join(person_folders[j], random.choice(folder2_images))))


In [42]:
# Assign labels to positive and negative pairs
positive_labelsc = [1] * len(positive_pairsc)
negative_labelsc = [0] * len(negative_pairsc)

print(negative_pairsc[0])
print(len(positive_pairsc), len(negative_pairsc))
# Combine positive and negative pairs and labels
custom_paired_images = positive_pairsc + negative_pairsc
custom_labels = positive_labelsc + negative_labelsc

('/kaggle/input/lfwdataset1/lfw/Custom1/2/Screenshot 2023-10-21 233210.png', '/kaggle/input/lfwdataset1/lfw/Custom1/1/IMG_20230317_223910.jpg')
2 1


In [44]:
# Validation dataset
validation_dataset = SiameseDataset(validation_paired_images, validation_labels, transforms=transform_list)
validation_loader = DataLoader(validation_dataset, batch_size=batch_size, shuffle=False)

# Testing dataset
test_dataset = SiameseDataset(test_paired_images, test_labels, transforms=transform_list)
test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)

#Custom dataset
custom_dataset = SiameseDataset(custom_paired_images, custom_labels, transforms=transform_list)
custom_loader = DataLoader(custom_dataset, batch_size=batch_size, shuffle=False)

Testing validation and test accuracies for model with ADAM optimizer and Cosine Annealing Learning rate scheduler

In [46]:
def evaluate(model, data_loader):
    model.eval()
    correct = 0
    total = 0
    with torch.no_grad():
        for data in data_loader:
            input1, input2, label = data
            input1, input2, label = input1.to(device), input2.to(device), label.to(device)
            output1, output2 = model(input1, input2)
            distances = F.pairwise_distance(output1, output2)
#             print(distances)
            predictions = distances < margin
            correct += (predictions == label.byte()).sum().item()
            total += label.size(0)
    accuracy = correct / total
    return accuracy

# Evaluate on validation and test datasets
validation_accuracy = evaluate(model, validation_loader)
test_accuracy = evaluate(model, test_loader)
# custom_accuracy = evaluate(model, custom_loader)

print(f"Validation Accuracy: {validation_accuracy:.4f}")
print(f"Test Accuracy: {test_accuracy:.4f}")
# print(f"Test Accuracy: {custom_accuracy:.4f}")


Validation Accuracy: 0.4691
Test Accuracy: 0.3282


In [48]:
# custom_accuracy = evaluate(model, custom_loader)
# print(f"Test Accuracy: {custom_accuracy:.4f}")

Train the model using ADAM optimizer and Reduce LR on plateau scheduler

In [50]:
from torch.optim.lr_scheduler import ExponentialLR

# Create the ExponentialLR scheduler with the desired gamma (exponential decay factor)
scheduler = ExponentialLR(optimizer_adam, gamma=0.95)  # Adjust gamma as needed

num_epochs = 1  # Adjust as needed

for epoch in range(num_epochs):
    for i, data in enumerate(train_loader, 0):
        input1, input2, label = data
        input1, input2, label = input1.to(device), input2.to(device), label.to(device)
        optimizer_adam.zero_grad()
        output1, output2 = model(input1, input2)
        loss = criterion(output1, output2, label, model)
        loss.backward()
        optimizer_adam.step()

        # Update the learning rate after each optimizer step
        scheduler.step()

        if (i + 1) % 10 == 0:
            print(f"Epoch [{epoch + 1}/{num_epochs}] Batch [{i + 1}/{len(train_loader)}] Loss: {loss.item():.4f}")


Epoch [1/1] Batch [10/5524] Loss: 0.7982
Epoch [1/1] Batch [20/5524] Loss: 0.7945
Epoch [1/1] Batch [30/5524] Loss: 0.6223
Epoch [1/1] Batch [40/5524] Loss: 0.6278
Epoch [1/1] Batch [50/5524] Loss: 0.7970
Epoch [1/1] Batch [60/5524] Loss: 0.7927
Epoch [1/1] Batch [70/5524] Loss: 0.4607
Epoch [1/1] Batch [80/5524] Loss: 1.1060
Epoch [1/1] Batch [90/5524] Loss: 0.4566
Epoch [1/1] Batch [100/5524] Loss: 0.7916
Epoch [1/1] Batch [110/5524] Loss: 1.1237
Epoch [1/1] Batch [120/5524] Loss: 0.7831
Epoch [1/1] Batch [130/5524] Loss: 0.4623
Epoch [1/1] Batch [140/5524] Loss: 0.9549
Epoch [1/1] Batch [150/5524] Loss: 1.6140
Epoch [1/1] Batch [160/5524] Loss: 0.7895
Epoch [1/1] Batch [170/5524] Loss: 0.2957
Epoch [1/1] Batch [180/5524] Loss: 0.4666
Epoch [1/1] Batch [190/5524] Loss: 0.3024
Epoch [1/1] Batch [200/5524] Loss: 0.6287
Epoch [1/1] Batch [210/5524] Loss: 0.7945
Epoch [1/1] Batch [220/5524] Loss: 0.2873
Epoch [1/1] Batch [230/5524] Loss: 0.6260
Epoch [1/1] Batch [240/5524] Loss: 1.2997
E

In [51]:
validation_accuracy = evaluate(model, validation_loader)
test_accuracy = evaluate(model, test_loader)
# custom_accuracy = evaluate(model, custom_loader)

print(f"Validation Accuracy: {validation_accuracy:.4f}")
print(f"Test Accuracy: {test_accuracy:.4f}")
# print(f"Test Accuracy: {custom_accuracy:.4f}")

Validation Accuracy: 0.4691
Test Accuracy: 0.3282


In [54]:
from torch.optim.lr_scheduler import ExponentialLR

optimizer_nadam = optim.NAdam(model.parameters(), lr=0.001)

# Create the ExponentialLR scheduler with the desired gamma (exponential decay factor)
scheduler = ExponentialLR(optimizer_nadam, gamma=0.95)  # Adjust gamma as needed

num_epochs = 2  # Adjust as needed

for epoch in range(num_epochs):
    for i, data in enumerate(train_loader, 0):
        input1, input2, label = data
        input1, input2, label = input1.to(device), input2.to(device), label.to(device)
        optimizer_nadam.zero_grad()
        output1, output2 = model(input1, input2)
        loss = criterion(output1, output2, label, model)
        loss.backward()
        optimizer_nadam.step()

        # Update the learning rate after each optimizer step
        scheduler.step()

        if (i + 1) % 10 == 0:
            print(f"Epoch [{epoch + 1}/{num_epochs}] Batch [{i + 1}/{len(train_loader)}] Loss: {loss.item():.4f}")


Epoch [1/2] Batch [10/5524] Loss: 0.8633
Epoch [1/2] Batch [20/5524] Loss: 0.3112
Epoch [1/2] Batch [30/5524] Loss: 0.6310
Epoch [1/2] Batch [40/5524] Loss: 0.6143
Epoch [1/2] Batch [50/5524] Loss: 0.4683
Epoch [1/2] Batch [60/5524] Loss: 0.4638
Epoch [1/2] Batch [70/5524] Loss: 1.4421
Epoch [1/2] Batch [80/5524] Loss: 0.9417
Epoch [1/2] Batch [90/5524] Loss: 1.0885
Epoch [1/2] Batch [100/5524] Loss: 0.7725
Epoch [1/2] Batch [110/5524] Loss: 0.7632
Epoch [1/2] Batch [120/5524] Loss: 0.4355
Epoch [1/2] Batch [130/5524] Loss: 0.9319
Epoch [1/2] Batch [140/5524] Loss: 0.9490
Epoch [1/2] Batch [150/5524] Loss: 0.6099
Epoch [1/2] Batch [160/5524] Loss: 0.6050
Epoch [1/2] Batch [170/5524] Loss: 1.1089
Epoch [1/2] Batch [180/5524] Loss: 0.9294
Epoch [1/2] Batch [190/5524] Loss: 0.6046
Epoch [1/2] Batch [200/5524] Loss: 0.4485
Epoch [1/2] Batch [210/5524] Loss: 0.6124
Epoch [1/2] Batch [220/5524] Loss: 1.3999
Epoch [1/2] Batch [230/5524] Loss: 0.9567
Epoch [1/2] Batch [240/5524] Loss: 1.0984
E

In [55]:
validation_accuracy = evaluate(model, validation_loader)
test_accuracy = evaluate(model, test_loader)

print(f"Validation Accuracy: {validation_accuracy:.4f}")
print(f"Test Accuracy: {test_accuracy:.4f}")

Validation Accuracy: 0.4691
Test Accuracy: 0.3282


Tested for Validation and test set

Testing for custom image input

In [57]:
custom_accuracy = evaluate(model, custom_loader)
print(f"Test Accuracy: {custom_accuracy:.4f}")